### ISSUE TITLE
Error Indexing cron: [pdc_discovery/staging] Net::ReadTimeout: Net::ReadTimeout with #<Socket:(closed)>

### ISSUE THREAD

**Issue by carolyncole** [pulibrary/pdc_discovery#485]
Notes document here: https://docs.google.com/document/d/1Rgc7KIlF9QVkPq8JHJl0JR-_kY7015O3dzde_RteGZk/edit#heading=h.zgv1l31zpjk7

## Backtrace

    line 126 of [PROJECT_ROOT]/lib/traject/solr_cloud_helper.rb: delete_collection!
    line 61 of [PROJECT_ROOT]/lib/traject/solr_cloud_helper.rb: collection_writer_for_alias
    line 36 of [PROJECT_ROOT]/lib/traject/solr_cloud_helper.rb: create_collection_writer

[View full backtrace and more info at honeybadger.io](https://app.honeybadger.io/projects/95072/faults/99601851)

**Comment by hectorcorrea** [comment]
The related HoneyBadger notice for production: https://app.honeybadger.io/projects/95072/faults/99595191

**Comment by hectorcorrea** [comment]
The culprit was that our Solr cluster in staging became unstable. 

The problem logged by our app was that it could not create a new Solr collection and in looking closer I noticed that we could not perform this activity from the Solr Web UI either. I restarted the Solr nodes via `sudo service solr restart` in staging (`lib-solr-staging4-5-6`) and the problem went away. Restarting the service in `lib-solr-staging6` took an abnormal amount of time (like 10 minutes) but eventually it restarted and was OK.

I reindexed in staging via `bundle exec rake index:research_data` and the reindex completed as normal.

**Comment by hectorcorrea** [comment]
Ran reindex in production to make sure that cluster was OK and reindex completed successfully.

**Comment by leefaisonr** [comment]
This is happening again. Bess and Robert are looking into it.

**Comment by bess** [comment]
During ansible office hours, @carolyncole suggested that the issue might be that one of the three solr nodes is in a different geographic location than the other two. @acozine is moving lib-solr-staging-6 to the same place as the other solr staging nodes and we hope this will reduce network latency and resolve the issue. Fingers crossed!

**Comment by bess** [comment]
We moved all three nodes to the same network, and then we moved all three zookeeper nodes also to the same network. Again, this seemed to resolve the problem for awhile, but it was back within 24 hours.

<img width="910" alt="Screenshot 2023-11-29 at 7 44 49 AM" src="https://github.com/pulibrary/pdc_discovery/assets/65608/1179e0eb-e015-4721-9f19-062a7fcb50aa">

**Comment by acozine** [comment]
Regarding the "which zookeepers" question - yesterday we moved several zookeeper machines, based on the config we found on the Solr8 boxes. Here's a list of the machines that are currently all in the same datacenter and should not encounter latency issues communicating with each other - I believe these are the correct ones (in other words, these three zookeeper boxes are connecting to these three solr boxes):

- lib-solr-staging4
- lib-solr-staging5
- lib-solr-staging6
- lib-zk-staging1
- lib-zk-staging2
- lib-zk-staging3

**Comment by hectorcorrea** [comment]
I noticed that our reindex task seems to hang while fetching the data from DataSpace...the job never finishes and then the schedule reindex kicks off 1/2 hour later and that's the error that we are seeing in Honeybadger, but the real problem was that the previous reindex did not complete. It looks to me like Solr is (understandably) blocking the second reindex. 

```
deploy@pdc-discovery-staging1:/opt/pdc_discovery/current$ grep "Indexing" log/staging.log
I, [2023-11-30T00:00:36.057221 #738840]  INFO -- : Indexing: Created a new collection for writing: http://lib-solr8-staging.princeton.edu:8983/solr/pdc-discovery-staging-1
I, [2023-11-30T00:00:36.057376 #738840]  INFO -- : Indexing: Fetching PDC Describe records
I, [2023-11-30T00:00:36.057775 #738840]  INFO -- : Indexing: Harvesting and indexing PDC Describe data started
I, [2023-11-30T00:06:59.830116 #738840]  INFO -- : Indexing: Harvesting and indexing PDC Describe data completed
I, [2023-11-30T00:06:59.830258 #738840]  INFO -- : Indexing: Fetching DataSpace records
I, [2023-11-30T00:06:59.830488 #738840]  INFO -- : Indexing: Harvesting and indexing DataSpace research data collections started

Notice that the DataSpace reindex started at 00:06:59 never finished.
A complete reindex will also log:

Indexing: Harvesting and indexing DataSpace research data collections completed
Indexing: Fetching completed
Indexing: Updated Solr to read from the new collection: pdc-discovery-staging/ -> ...

and then the reindex kicked in at 00:30:03 but I guess since the previous one has not finished Solr is blocking our second reindex.

```

I've deployed a version of Discovery (PR https://github.com/pulibrary/pdc_discovery/pull/538) that does not fetch DataSpace data to see if our reindex runs without issues for 24 hours or more.

**Comment by hectorcorrea** [comment]
Disabling the DataSpace reindex did not fix the issue. I updated the post incident document (https://docs.google.com/document/d/1Rgc7KIlF9QVkPq8JHJl0JR-_kY7015O3dzde_RteGZk/edit) with my finding from today.

I noticed that although the commands that we issue to recreate the index (DELETE, CREATE) timed out and look like they did not succeeded, *as soon as I restarted one of the Solr nodes* they look like they actually succeeded 😱 -- i.e. the delete collection was gone and the created collection showed up in the Solr dashboard.
